From f911f45786b0ad2b3ce9780c819eb2a2599d5a33 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E9=BB=84=E6=B6=9B?= <huangtao@rock-chips.com>
Date: Sat, 30 Jul 2011 22:16:37 +0800
Subject: [PATCH] Revert "[ARM] Do not call flush_cache_user_range with
 mmap_sem held"

This reverts commit 7bc14acfbb86b00a93ddb1c40d9d2050cd7188fe.
---
 arch/arm/include/asm/cacheflush.h | 2 +-
 arch/arm/kernel/traps.c           | 4 +---
 2 files changed, 2 insertions(+), 4 deletions(-)

diff --git a/arch/arm/include/asm/cacheflush.h b/arch/arm/include/asm/cacheflush.h
index 3e32991c04fa..3d0cdd21b882 100644
--- a/arch/arm/include/asm/cacheflush.h
+++ b/arch/arm/include/asm/cacheflush.h
@@ -381,7 +381,7 @@ extern void flush_ptrace_access(struct vm_area_struct *vma, struct page *page,
  * Harvard caches are synchronised for the user space address range.
  * This is used for the ARM private sys_cacheflush system call.
  */
-#define flush_cache_user_range(start,end) \
+#define flush_cache_user_range(vma,start,end) \
 	__cpuc_coherent_user_range((start) & PAGE_MASK, PAGE_ALIGN(end))
 
 /*
diff --git a/arch/arm/kernel/traps.c b/arch/arm/kernel/traps.c
index 14abab76019a..c1750dd70ab6 100644
--- a/arch/arm/kernel/traps.c
+++ b/arch/arm/kernel/traps.c
@@ -438,9 +438,7 @@ do_cache_op(unsigned long start, unsigned long end, int flags)
 		if (end > vma->vm_end)
 			end = vma->vm_end;
 
-		up_read(&mm->mmap_sem);
-		flush_cache_user_range(start, end);
-		return;
+		flush_cache_user_range(vma, start, end);
 	}
 	up_read(&mm->mmap_sem);
 }
-- 
2.35.3

